src/history/hs.c

```diff
@@ -824,179 +823,166 @@ __wt_find_hs_upd(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE **upd
-                __wt_free_update_list(session, &mod_upd);
-                mod_upd = NULL;
-            }
-            /* After we're done looping over modifies, reset the read timestamp. */
-            __hs_restore_read_timestamp(session, saved_timestamp);
-            WT_STAT_CONN_INCR(session, cache_hs_read_squash);
+            WT_ERR(hs_cursor->get_value(hs_cursor, &hs_stop_ts_tmp, &durable_timestamp_tmp,
+              &prepare_state_tmp, &upd_type, hs_value));
         }
 
-        /* Allocate an update structure for the record found. */
-        WT_ERR(__wt_update_alloc(session, hs_value, &upd, &size, upd_type));
-        upd->txnid = hs_start.txnid;
-        upd->durable_ts = durable_timestamp;
-        upd->start_ts = hs_start.timestamp;
-        upd->prepare_state = prepare_state;
+        WT_ASSERT(session, upd_type == WT_UPDATE_STANDARD);
+        while (modifies.size > 0) {
+            __wt_modify_vector_pop(&modifies, &mod_upd);
+            WT_ERR(__wt_modify_apply_item(session, hs_value, mod_upd->data, false));
+            __wt_free_update_list(session, &mod_upd);
+            mod_upd = NULL;
+        }
+        /* After we're done looping over modifies, reset the read timestamp. */
+        __hs_restore_read_timestamp(session, saved_timestamp);
+        WT_STAT_CONN_INCR(session, cache_hs_read_squash);
+    }
 
-        /*
-         * When we find a prepared update in the history store, we should add it to our update list
-         * and subsequently delete the corresponding history store entry. If it gets committed, the
-         * timestamp in the las key may differ so it's easier if we get rid of it now and rewrite
-         * the entry on eviction/commit/rollback.
-         *
-         * FIXME-PM-1521: review the code in future
-         */
-        if (prepare_state == WT_PREPARE_INPROGRESS) {
-            WT_ASSERT(session, !modify);
-            switch (cbt->ref->page->type) {
-            case WT_PAGE_COL_FIX:
-            case WT_PAGE_COL_VAR:
-                recnop = hs_key->data;
-                WT_ERR(__wt_vunpack_uint(&recnop, 0, &recno));
-                WT_ERR(__wt_col_modify(cbt, recno, NULL, upd, WT_UPDATE_STANDARD, false));
-                break;
-            case WT_PAGE_ROW_LEAF:
-                WT_ERR(__wt_row_modify(cbt, hs_key, NULL, upd, WT_UPDATE_STANDARD, false));
-                break;
-            }
+    /* Allocate an update structure for the record found. */
+    WT_ERR(__wt_update_alloc(session, hs_value, &upd, &size, upd_type));
+    upd->txnid = WT_TXN_NONE;
+    upd->durable_ts = durable_timestamp;
+    upd->start_ts = hs_start_ts;
+    upd->prepare_state = prepare_state;
 
-            ret = hs_cursor->remove(hs_cursor);
-            if (ret != 0)
-                WT_PANIC_ERR(session, ret,
-                  "initialized prepared update but was unable to remove the corresponding entry "
-                  "from hs");
+    /*
```

Comment:
same as above, we should never have a prepared update in history store. We could get rid of prepared state from the value part of the entry in history store.